bip-taro.mediawiki

```diff
@@ -219,21 +225,37 @@ Taro-specific commitments, then the same or different leaf version can be used
 to gate verification of the new behavior.
 
 An <code>asset_id</code> is the 32-byte hash of: 
-* <code>sha256(genesis_outpoint || asset_tag || asset_meta)</code>
+* <code>sha256(genesis_outpoint || asset_tag || asset_meta || output_position)</code>
 
 where:
 * <code>genesis_outpoint</code> is the first previous input outpoint used in the asset genesis transaction serialized in Bitcoin wire format.
 * <code>asset_tag</code> is a random 32-byte value that represents a given asset, and can be used to link a series of discrete assets into a single asset family. In practice, this will typically be the hash of a human readable asset name.
 * <code>asset_meta</code> is an opaque 32-byte value that can be used to commit to various metadata including external links, documents, stats, attributes, images, etc. Importantly, this field is considered to be ''immutable''.
+* <code>output_position</code> is the output position of the output that caries the unique Taro commitment in the genesis transaction. 
 
 Given the above structure, the <code>asset_id</code> is guaranteed to be
 globally unique from the PoV of the chain, as thanks to
 [[../master/bip-0034.mediawiki|BIP 34]]
 as outpoints can never repeat once serialized within the block chain.
 
 In addition, we enforce a rule that a given output MUST only have a ''single''
-root Taro asset commitment (any other duplicates cannot be spent due to the way
-asset witnesses bind to the input outpoint). 
+root Taro asset commitment. In order to verify this property, we enforce the
+following constraints on the position of the root Taro asset tree commitment:
+* The asset root commitment MUST be the leftmost or rightmost leaf in the tapscript tree:
+** We enforce this by requiring the reveal of the taro commitment in the script tree to either be of depth 0 or 1.
+* For a reveal proof of depth 0, the Taro commitment is the only element in the control block proof, as it directly becomes the root hash.
+* For a reveal proof of depth 1, we require the prover to also reveal the pre-image of the sibling hash.
+** With a root position of depth 1, the Taro commitment is either the leftmost or rightmost element of the fully sorted tapscript tree.
```

Comment:
IIUC, if we always require a depth of 1, then we'll need to hash all the script leaves first until we arrive at a single branch, and then hash that with the Taro commitment leaf? This would resemble the following tree:

```
             tapscript root
               /        \
              /          \
  taro commitment leaf    tap branch
                        /  \       /  \
                       /    \     /    \
                     leaf-0 ...  ... leaf-n
```

Reply:
That's a correct diagram, but consider that that "tap branch" there can actually be an independent tree, with us just merging that tree with our new final leaf. This makes it easy for wallets as they can construct their main tree however they want, and then the taro libraries can just merge things into place. 

At this point, I think it may be good to also roll out some structure taproot commitment stuff into its own BIP. As with this structure, you can naturally define unique locations for other systems within that "left" sub-tree. Given that the leaf itself also isn't bound by the normal taproot "sort then hash", we don't lose ordering information, so we can obtain particular placements. 

I'm working on an update here that adds some psueod code, but in this case we'd get a control block that's 65 bytes (the internal key, and the single tap branch). From here'd wed ask for that pre-image (the sorted concat in the leaf case), and then assert that it's an actually branch. When this is the internal portion of the tree (past the leaves), we'll always just have two items there to sort and hash, and the tagged hash construct lets us ensure that they're actually branches. In other words we don't actually need to know what actually lies below, as the control block must always be 65 or 33 bytes (33 is just the single level).

Reply:
So to enumerate the cases a bit more with the depth of position 1:
  * It's a leaf, we use the tagged hash for leafs to confirm, and then confirm that isn't a taro commitment 
  * It's a branch:
     * It's an internal branch, we used the tagged hash for branches to confirm, gg 
     * It's a branch composed of two leaves, we used the tagged hash to confirm, gg
    
This will need to be reflected in the proof format, and also elsewhere in the Universe structure.

Reply:
So we want a control block proof for the taro commitment leaf, normally we'd just get that sibling tap branch, but we also want it's pre-image as well (I think this is what you were actually asking).

Reply:
> It's an internal branch, we used the tagged hash for branches to confirm, gg
It's a branch composed of two leaves, we used the tagged hash to confirm, gg

How are these two cases distinct? In either case, we just need to reveal the child hashes to obtain the tap branch hash.

Reply:
True we can make a simplification here, since we don't actually need to go an additional level there.